clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -761,48 +762,79 @@ static CompoundStmt *CreateOpenCLKernelBody(Sema &S,
       // Recursively search for accessor fields to initialize them with kernel
       // parameters
       std::function<void(const CXXRecordDecl *, Expr *)>
```

Comment:
please make this NOT std::function.  I realize this is existing, but this is a lambda, the type erasure cost of a std::function here is pretty horrible.  Just replace the type above with 'auto'.

Reply:
The suggested change fails to compile because there is a recursive call to getExprForWrappedAccessorInit, and because of that, "auto" cannot be used.
I would rather not get into making this change since it is in existing code.

Reply:
Ah, missed that it is calling itself.  That said,you ARE modifying everything else with it, so it is completely within the scope of this change.

Reply:
When auto is used the compiler does not accept the call within the body, because the type deduction hasn't happened yet. Can you suggest a way to write the type of this lambda, that doesn't use auto, and also doesn't use std::function?

Reply:
You cannot name the type of a lambda as anything but 'auto'.

You cant use lambdas recursively unfortunately, the type of the lambda isn't complete until after the closing brace.  SO the only alternative is a function object.    In this case, I'm OK with std::function for now.  Normally, it would be in the scope of the review, but I don't think its worth changing since you cannot just use the lambda's type.